### PR TITLE
Support cuComplex.h in `cupy.RawKernel` and `cupy.RawModule`

### LINKED ISSUE #1867: easier way to query device attributes?
I was curious if there is interest in adding a more user-friendly way to retrieve device attributes such as MAX_THREADS_PER_BLOCK, MAX_GRID_DIM_X, etc. In CUDA 9.2/CUDA 10 there are 101 such attributes defined in `cuda.h`. 

It seems that these can already be retrieved via `cupy.cuda.runtime.deviceGetAttribute`, but one needs to know the appropriate enum value for each attribute of interest.

Is there any interest in adding a function that would return the values of all available attributes as a dict or other similar structure that is more user-friendly?

A basic solution I came up with was to copy enum values from `cuda.h` into one of the cython files like `runtime.pxd`.  If there is interest I can make a PR.

e.g.
```cython
cpdef enum:
    CU_DEVICE_ATTRIBUTE_MAX_THREADS_PER_BLOCK = 1              # Maximum number of threads per block
    CU_DEVICE_ATTRIBUTE_MAX_BLOCK_DIM_X = 2                    # Maximum block dimension X
    CU_DEVICE_ATTRIBUTE_MAX_BLOCK_DIM_Y = 3                    # Maximum block dimension Y
    CU_DEVICE_ATTRIBUTE_MAX_BLOCK_DIM_Z = 4                    # Maximum block dimension Z
    CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_X = 5                     # Maximum grid dimension X
    CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_Y = 6                     # Maximum grid dimension Y
    CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_Z = 7                     # Maximum grid dimension Z
    CU_DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_BLOCK = 8        # Maximum shared memory available per block in bytes
```
etc.

Then a convenience function similar to the following would create a `dict` with attribute names as the keys.
```python

def get_attributes_dict(device_id):
    """Return a dict containing all device attributes."""
    d = {}
    for k, v in runtime.__dict__.items():
        if 'CU_DEVICE_ATTRIBUTE' in k:
            try:
                name = k.replace('CU_DEVICE_ATTRIBUTE_', '')
                d[name] = runtime.deviceGetAttribute(v, device_id)
            except runtime.CUDARuntimeError:
                pass
    return d
```
(the except case is necessary to skip any attributes that were undefined in older CUDA versions)

The output of the function is a dict such as:
```python
{ 'MAX_THREADS_PER_BLOCK': 1024,
  'MAX_BLOCK_DIM_X': 1024,
  'MAX_BLOCK_DIM_Y': 1024,
  'MAX_BLOCK_DIM_Z': 64,
  'MAX_GRID_DIM_X': 2147483647,
  'MAX_GRID_DIM_Y': 65535,
  'MAX_GRID_DIM_Z': 65535,
  'MAX_SHARED_MEMORY_PER_BLOCK': 49152,
...

### LINKED ISSUE #2111: Use complex number in cupy.RawKernel
Hi!

I wonder what is the native/suggested way to use complex numbers in Cupy.RawKernel?

I was using `#include<cuComplex.h>` until I got an error on my newly installed cupy on another computer. I scrambled through your code and it seems to me that the preferred way is to use `#include<cupy/complex/complex.h>`. Is this the preferred library to use for cupy?

### PR DESCRIPTION
**Pull request by leofang**
Closes #1866. Closes #2111.

Previously the support of complex numbers in `cupy.RawKernel` (and thus `cupy.RawModule`) was poorly documented, see https://github.com/cupy/cupy/pull/1866#issuecomment-457791608 and #2111 for example. This PR aims to solve this problem once and for all. This PR supersedes #1866.

The support of the APIs from `cuComplex.h` is enabled by prepending a set of C macros that replace those APIs by their Thrust counterparts. Thus, users can just copy and paste their `cuComplex`-based codes to `cupy.RawKernel` or `cupy.RawModule`, and make them work without any modification. Note that the complex mathematical functions are supported without any additional macros.

For backward compatibility, the newly added option `enable_cuComplex` is set to `False` by default.

attn: @grlee77

### CODE REVIEW

**Comment by leofang** [comment]
Note that the commits in #1866 are taken over to here.

File: cupy/core/carray.pxi [hunk 1]
```diff
@@ -122,9 +122,60 @@ cpdef str _get_header_source():
...
+            break
+    source = '\n'.join(source_lines)
+
+    # Second, we use macros to replace cuComplex APIs by those of Thrust's
+    macro = r'''
+    /* ------------------- single complex ------------------- */
+    #define cuFloatComplex                complex<float>
+    #define cuComplex                     complex<float>
+    #define cuCrealf                      real
+    #define cuCimagf                      imag
```
Comment (leofang) [inline cupy/core/carray.pxi]: This is enabled by #2520.

File: cupy/core/carray.pxi [hunk 2]
```diff
@@ -122,9 +122,60 @@ cpdef str _get_header_source():
...
+    #define cuCmulf(A, B)                 A * B
+    #define cuCdivf(A, B)                 A / B
+    #define cuCabsf                       abs
+    #define cuComplexDoubleToFloat        complex<float>
+    #define cuCfmaf(A, B, C)              A * B + C
+
+    /* ------------------- double complex ------------------- */
+    #define cuDoubleComplex               complex<double>
+    #define cuCreal                       real
+    #define cuCimag                       imag
```
Comment (leofang) [inline cupy/core/carray.pxi]: Ditto.

**Comment by hvy** [comment]
Sorry for the drop by comment but thanks for the PR @leofang. @kmaehashi could you have a look at this PR? If you're busy, then @asi1024 said he might be able to take over.

**Comment by leofang** [comment]
@hvy No need to apologize! Thanks for your followup and kind reply.

File: cupy/core/carray.pxi [hunk 3]
```diff
@@ -122,9 +122,60 @@ cpdef str _get_header_source():
     return _header_source
 
 
+cdef str _convert_cuComplex_to_Thrust(str source):
+    cdef str macro, line
+    cdef list source_lines = [line for line in source.split('\n')]
+
+    # First, we comment out the line that includes cuComplex.h
+    for i, line in enumerate(source_lines):
+        if '#include' in line and 'cuComplex.h' in line:
```
Comment (kmaehashi) [inline cupy/core/carray.pxi]: Well, this approach sounds fragile and may surprise users. Do we really need to support `cuComplex.h`? How about providing a bridge interface in the separate header file (in that case users need to replace `cuComplex.h` to sth like `cupy/cuComplex_bridge.h`)?

Comment (leofang) [reply cupy/core/carray.pxi]: Thanks for raising your concern, @kmaehashi.

> Do we really need to support `cuComplex.h`?

Short answer: Yes. Please do. 

Long answer: #2111 is just a long list of the cases I have encountered that request this. Please do not surprise our users. Many scientific projects need complex numbers, but some existing CUDA projects either do not want to depend on Thrust, or it's too large to do a clean overhaul to transit to Thrust's complex types. So, when people port code to CuPy, they are often frustrated. I've helped many people resolve this, and I think this PR solves the issue once and for all. (So people would stop bothering their local CuPy expert, i.e., me...)

If you are really, really against such a support, can you please at least merge #1866? It's hanging there for quite a while, and it's really frustrating. In either case, we need to make it clear to the users. No surprises.

> How about providing a bridge interface in the separate header file (in that case users need to replace `cuComplex.h` to sth like `cupy/cuComplex_bridge.h`)?

I do not think this is less fragile than the current Pythonic approach. First, this requires code change (although it's only one line), which is against the design principle I had in mind when pushing for `RawModule`. (Users should just import the source code and let CuPy take care of the rest to make it work.)

Second, in the end the macros in this PR go into `cupy/cuComplex_bridge.h`, but the net effect, after the C preprocessing, is identical. So I don't see why we wanna bother users to make a change?

If you are still concerned and not fully convinced, may I propose to mark the new `enable_cuComplex` flag as experimental and subject to change? Let me reiterate that the changes in this PR are not enabled by default, so normal use cases would just continue to work (as you can see from the Jenkins). It would not hurt to consider this PR.

Thanks again, @kmaehashi.

Comment (leofang) [reply cupy/core/carray.pxi]: @kmaehashi just check in. Any comment? Thanks.

Comment (kmaehashi) [reply cupy/core/carray.pxi]: OK, now I understand your motivation to use cuComplex code directly on CuPy RawModule/RawKernel.

* Regarding the option name, how about `translate_cucomplex` instead of `enable_cuComplex`? (1) It is difficult for users to imagine that the code is automatically rewritten when `enable` is set. (2) pep8 recommends lowercase for variable names.
* Could you make the option a keyword-only argument so that we can easily consider reorder/renaming later?
* Please use more stricter regex that matches the whole line to detect a header file. You need to consider cases like: `# include <cuComplex.h>` or `// #include <cuComplex.h>`
* Could you extract `macro` contents as a separate file (`cupy/cuComplex_bridge.h`)?
  Then replace `#include <cuComplex.h>` with `#include <cupy/cuComplex_bridge.h>  // translate_cucomplex` to minimize the amount of modification? (this also keeps the original line numbers so it's better for cuda-gdb)

Comment (leofang) [reply cupy/core/carray.pxi]: Thank you very much for your understanding and detailed comments, @kmaehashi. I'm working on it. 

Two quick replies:
> Could you make the option a keyword-only argument so that we can easily consider reorder/renaming later?

If I understand you correctly, I don't think for `compile_with_cache` this is possible unfortunately. `cdef` and `cpdef` functions do not support `**kwargs` in the end of the function signature (will get an `Expected ')', found '**'` error).

> this also keeps the original line numbers so it's better for cuda-gdb

This is a very good suggestion which I would never come up myself! I never used `cuda-gdb` after all...

**Comment by leofang** [comment]
Done. PTAL.

File: cupy/core/carray.pxi (hunk 3 above)
Comment (kmaehashi) [reply cupy/core/carray.pxi]: > If I understand you correctly, I don't think for compile_with_cache this is possible unfortunately. cdef and cpdef functions do not support **kwargs in the end of the function signature (will get an Expected ')', found '**' error).

Yes, you're correct. I think it's OK for `compile_with_cache` as it is not public (documented) interface.

File: cupy/core/carray.pxi [hunk 4]
```diff
@@ -122,9 +123,21 @@ cpdef str _get_header_source():
     return _header_source
 
 
+cdef inline str _convert_cuComplex_to_Thrust(str source):
+    source = re.sub(r'<cuComplex.h>',
```
Comment (kmaehashi) [inline cupy/core/carray.pxi]: Ah, what I meant by "regex that matches the whole line" was like this:

```py
# module level (to precompile the regex)
cucomplex_include_tokens = ['', '#', 'include', '<', r'cuComplex\.h', '>', '']
cucomplex_include_pattern = re.compile(r'\s*'.join(cucomplex_include_tokens))

# in _convert_cuComplex_to_Thrust
lines = []
for line in source.splitlines(keepends=True):
    if cucomplex_include_pattern.match(line):
        lines += '#include <cupy/cuComplex_bridge.h>  // translate_cucomplex'
    else:
        lines += line
return ''.join(lines)
```

File: cupy/core/raw.pxd [hunk 5]
```diff
@@ -6,6 +6,7 @@ cdef class RawKernel:
         readonly tuple options
         object _kernel
         readonly str backend
+        bint cuComplex
```
Comment (kmaehashi) [inline cupy/core/raw.pxd]: nitpick: use `translate_cucomplex` for consistency and PEP8

File: cupy/core/raw.pyx [hunk 6]
```diff
@@ -24,9 +24,12 @@ cdef class RawKernel:
         options (str): Compiler options passed to NVRTC. For details, see
             https://docs.nvidia.com/cuda/nvrtc/index.html#group__options.
         backend (str): Either `nvrtc` or `nvcc`. Defaults to `nvrtc`
+        translate_cucomplex (bool): Whether the CUDA source includes the
+            "cuComplex.h" header or not. Defaults to False.
```
Comment (kmaehashi) [inline cupy/core/raw.pyx]: Please describe a behavior a bit more so that users can understand that the code is modified by CuPy. I want avoid surprising users 😉

File: cupy/core/raw.pyx [hunk 7]
```diff
@@ -24,9 +24,12 @@ cdef class RawKernel:
...
     """
 
-    def __init__(self, code, name, options=(), backend='nvrtc'):
+    def __init__(self, code, name, options=(), backend='nvrtc',
+                 translate_cucomplex=False):
```
Comment (kmaehashi) [inline cupy/core/raw.pyx]: Please make it keyword-only.

File: cupy/core/raw.pyx [hunk 8]
```diff
@@ -210,11 +216,14 @@ cdef class RawModule:
             needed. For details, see
             https://docs.nvidia.com/cuda/nvrtc/index.html#group__options.
         backend (str): Either `nvrtc` or `nvcc`. Defaults to `nvrtc`
+        translate_cucomplex (bool): Whether the CUDA source includes the
+            "cuComplex.h" header or not. Defaults to False.
```
Comment (kmaehashi) [inline cupy/core/raw.pyx]: ditto.

File: cupy/core/raw.pyx [hunk 9]
```diff
@@ -210,11 +216,14 @@ cdef class RawModule:
...
 
     .. note::
         Each kernel in ``RawModule`` possesses independent function attributes.
     """
-    def __init__(self, code_or_path, options=(), backend='nvrtc'):
+    def __init__(self, code_or_path, options=(), backend='nvrtc',
+                 translate_cucomplex=False):
```
Comment (kmaehashi) [inline cupy/core/raw.pyx]: ditto.

File: cupy/core/include/cupy/cuComplex_bridge.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+#define cuDoubleComplex               complex<double>
+#define cuCreal                       real
+#define cuCimag                       imag
+#define make_cuDoubleComplex(A, B)    complex<double>(A, B)
+#define cuConj                        conj
+#define cuCadd(A, B)                  (A + B)
+#define cuCsub(A, B)                  (A - B)
+#define cuCmul(A, B)                  (A * B)
+#define cuCdiv(A, B)                  (A / B)
+#define cuCabs                        abs
```
Comment (kmaehashi) [inline cupy/core/include/cupy/cuComplex_bridge.h]: This seems uncovered by test. Is it difficult to cover all functions?
(I haven't checked if all functions are covered in the test code; this is just one I noticed)

File: cupy/core/carray.pxi (hunk 4 above)
Comment (leofang) [reply cupy/core/carray.pxi]: I see. Sorry I know nothing about regex. While your regex code looks fine to me, for some reason it can't detect the following cases:
```python
>>> cucomplex_include_pattern.match(r'#include <cuComplex.h>')  # ok
<_sre.SRE_Match object; span=(0, 22), match='#include <cuComplex.h>'>
>>> cucomplex_include_pattern.match(r'//#include <cuComplex.h>')  # miss!
>>> cucomplex_include_pattern.match(r'// #include <cuComplex.h>')  # miss!
>>> cucomplex_include_pattern.match(r'// # include <cuComplex.h> ')  #miss!
```
@kmaehashi could you help me diagnose please?

File: cupy/core/raw.pxd (hunk 5 above)
Comment (leofang) [reply cupy/core/raw.pxd]: done

File: cupy/core/raw.pyx (hunk 6 above)
Comment (leofang) [reply cupy/core/raw.pyx]: Haha OK I agree. This is done.

File: cupy/core/raw.pyx (hunk 7 above)
Comment (leofang) [reply cupy/core/raw.pyx]: Done.

File: cupy/core/include/cupy/cuComplex_bridge.h (hunk 10 above)
Comment (leofang) [reply cupy/core/include/cupy/cuComplex_bridge.h]: Not difficult. I was just lazy. I think I cover all functions now.

**Comment by leofang** [comment]
The requested changes are done except for the regex one. Tested locally.

File: cupy/core/carray.pxi (hunk 4 above)
Comment (kmaehashi) [reply cupy/core/carray.pxi]:
```py
>>> cucomplex_include_pattern.match(r'//#include <cuComplex.h>')  # miss!
>>> cucomplex_include_pattern.match(r'// #include <cuComplex.h>')  # miss!
>>> cucomplex_include_pattern.match(r'// # include <cuComplex.h> ')  #miss!
```

This behavior is what I intended. They are commented out so we don't have to replace it.

Comment (leofang) [reply cupy/core/carray.pxi]: Oh oh, I see! I misunderstood what you meant. Will make a change before going to bed :)

File: cupy/core/carray.pxi [hunk 11]
```diff
@@ -122,9 +123,21 @@ cpdef str _get_header_source():
...
+
+
 cpdef function.Module compile_with_cache(
         str source, tuple options=(), arch=None, cachd_dir=None,
-        prepend_cupy_headers=True, backend='nvrtc'):
+        prepend_cupy_headers=True, backend='nvrtc', translate_cucomplex=False):
+    if translate_cucomplex:
+        source = _convert_cuComplex_to_Thrust(source)
+        _cupy_header_list.append('cupy/cuComplex_bridge.h')
+        prepend_cupy_headers=True
```
Comment (kmaehashi) [inline cupy/core/carray.pxi]: nitpick
```suggestion
        prepend_cupy_headers = True
```

File: cupy/core/raw.pyx [hunk 12]
```diff
@@ -24,20 +22,21 @@ cdef class RawKernel:
...
     """
 
-    def __init__(self, code, name, options=(), backend='nvrtc'):
-        if isinstance(code, six.binary_type):
-            code = code.decode('UTF-8')
-        if isinstance(name, six.binary_type):
-            name = name.decode('UTF-8')
-        if isinstance(backend, six.binary_type):
-            backend = backend.decode('UTF-8')
+    def __init__(self, code, name, options=(), backend='nvrtc', **kwargs):
```
Comment (kmaehashi) [inline cupy/core/raw.pyx]:
```suggestion
    def __init__(self, code, name, options=(), backend='nvrtc', *, translate_cucomplex=False):
```

Then you can remove the line below.

File: cupy/core/carray.pxi [hunk 13]
```diff
@@ -122,9 +123,21 @@ cpdef str _get_header_source():
     return _header_source
 
 
+cdef inline str _convert_cuComplex_to_Thrust(str source):
```
Comment (kmaehashi) [inline cupy/core/carray.pxi]: For consistency and PEP8
```suggestion
cdef inline str _translate_cucomplex_to_thrust(str source):
```

File: cupy/core/carray.pxi [hunk 14]
```diff
@@ -122,9 +123,21 @@ cpdef str _get_header_source():
...
+                    source)
+    return source
+
+
 cpdef function.Module compile_with_cache(
         str source, tuple options=(), arch=None, cachd_dir=None,
-        prepend_cupy_headers=True, backend='nvrtc'):
+        prepend_cupy_headers=True, backend='nvrtc', translate_cucomplex=False):
+    if translate_cucomplex:
+        source = _convert_cuComplex_to_Thrust(source)
```
Comment (kmaehashi) [inline cupy/core/carray.pxi]:
```suggestion
        source = _translate_cucomplex_to_thrust(source)
```

File: cupy/core/raw.pyx (hunk 12 above)
Comment (leofang) [reply cupy/core/raw.pyx]: Wait, I thought using `**kwargs` was what you wanted? Did I misunderstand?

File: cupy/core/carray.pxi (hunk 4 above)
Comment (leofang) [reply cupy/core/carray.pxi]: @kmaehashi This is added now.

**Comment by leofang** [comment]
@kmaehashi Most requested changes are done except for the `**kwargs` comment. 

Also, I took the liberty of removing `six`-related code in `cupy/core/raw.pyx`. Since PY2 support is dropped, I hope this is acceptable. If not, I'll bring them back. 

Thanks.

File: cupy/core/raw.pyx (hunk 12 above)
Comment (kmaehashi) [reply cupy/core/raw.pyx]: As in my original comment,

> Could you make the option a keyword-only argument so that we can easily consider reorder/renaming later?

I think [keyword-only arguments](https://python-3-for-scientists.readthedocs.io/en/latest/python3_advanced.html) (i.e., arguments list after `, * ,`) is enough. I just wanted to avoid adding new positional arguments.

File: cupy/core/raw.pyx [hunk 15]
```diff
@@ -24,20 +22,21 @@ cdef class RawKernel:
...
             https://docs.nvidia.com/cuda/nvrtc/index.html#group__options.
         backend (str): Either `nvrtc` or `nvcc`. Defaults to `nvrtc`
+        translate_cucomplex (bool): Whether the CUDA source includes the header
+            `cuComplex.h` or not. If set to ``True``, any code that uses the
+            functions from `cuComplex.h` will be translated to its Thrust
+            counterpart. Defaults to ``False``.
     """
 
-    def __init__(self, code, name, options=(), backend='nvrtc'):
-        if isinstance(code, six.binary_type):
```
Comment (kmaehashi) [inline cupy/core/raw.pyx]: If code is `bytes` type, it should be decoded even in Python 3. So these code cannot be removed.
(You can rewrite `six.binary_type` with `bytes`, but preferably do this in a separate PR as this is a different topic.)

Comment (leofang) [reply cupy/core/raw.pyx]: Didn't realize this 😢 I thought `code` is always a `str`?

**Comment by kmaehashi** [comment]
pfnCI, test this please.
(Running tests as this PR is almost complete.)

File: cupy/core/raw.pyx (hunk 12 above)
Comment (leofang) [reply cupy/core/raw.pyx]: Thanks, @kmaehashi! You're right. Apparently I need some rest 😞 Will resume to address the two changes tomorrow. I'd like to see the CI outcome anyway.

File: cupy/core/raw.pyx (hunk 15 above)
Comment (kmaehashi) [reply cupy/core/raw.pyx]: Generally we can assume that, but there's no reason to aggressively break backward compatibility. For example users may using binary stream to read file, e.g. `code=open('file.cu', 'rb').read()`.

Comment (leofang) [reply cupy/core/raw.pyx]: OK. But that's a really atypical/non-pythonic usage for reading a text-based file...

**Comment by leofang** [comment]
Everything seems to work. The only CI failure seems unrelated. The `six` support is restored, and the keyword-only argument is used. @kmaehashi PTAL, thanks.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by kmaehashi** [comment]
Test failures are not related to this PR.

**Comment by leofang** [comment]
Thanks a lot @kmaehashi for your help and @grlee77 for the update in the docs!
